### DISCUSSION TITLE
Refactor native Tradfri integration to use a data coordinator

### DISCUSSION THREAD

**Discussion by ggravlingen** [home-assistant/architecture#699] [Entity Models] [answered]
The native Ikea Tradfri integration, ie the one that uses IKEA's own gateway, was added to Home Assistant by myself and @balloob  around 2016 if I’m not mistaken. The official HA statistics suggests it’s currently being used by about 3% of Home Assistant’s user base. One of them is me after all these years.

There have been many reports of stability issues, where users are reporting that the connection to the gateway is lost and that a restart or reload is required every now and then. There have also been requests to add sensor data that is available in the pytradfri API but not exposed through Home Assistant for architectural reasons that I have discussed in the past with @MartinHjelmare. Many attempts have also been made developers here to solve the stability issues but it seems they are still present for some, but not all, users.

I discussed these issues a few years ago with @ludeeus, who pointed me to [this](https://developers.home-assistant.io/docs/integration_fetching_data/#coordinated-single-api-poll-for-data-for-all-entities) section of the documentation. I found the idea intriguing but haven’t had much time to investigate it. Until now.

Having analysed it, and built a working implementation as a custom integration, I propose we move from today’s subscription based solution to a polling based solution, using the coordinator approach.

The way I see it, there are a few possible advantages to this approach:
- ~~Stability: as we poll the gateway at a certain interval the load against it would be lower, hopefully making the integration more stable.~~ Edit: polling, as opposite to subscriptions, causes severe stability issues. We shouldn't poll.
- Sensor data: we could easily expose data users have been asking about, eg the battery level of the blind controller and the air quality information from the air purifier.

There are also a few disadvantages:
- Updates are no longer live: as we use a polling technique, the GUI updates are not instant if a light is turned on by an automation (it is instant if a user uses the GUI). For users where instant updates are important, perhaps the ZHA- or Conbee integrations is a better solution?
- Breaking change: when I loaded this as a custom integration, duplicates were created. I had to remove the integration and add it back again for all devices to show up again as expected. I’m thinking this is something I did wrong in coding and that this could be worked out in peer review.
- ~~Logs are chatty: every time the gateway is polled, the following log entry is created: `2022-01-01 13:44:52 INFO (MainThread) [coap] No DTLS connection active to (192.168.15.18, 5684, b'bdc3810925914f929ce38d231804909e'), creating one`. This is a log message created by the `aiocoap` library.~~ Edit: this stopped when I switched to a subscription based approach.

I have built a working version of this proposed change (including correctly exposing battery levels and air quality levels as sensors) and uploaded to a [Github repo](https://github.com/ggravlingen/tradfri_custom_component). I'm currently running it in my production environment to test how well it works. I’m happy to make the repo public if it would facilitate the discussions here, and understanding for the solution.

Edit: made the repo public.

**Comment by ggravlingen** [comment]
There are a few open issues on stability of the gateway in the `core` repo. We could probably find a group of brave beta testers of the proposed change in one of those issues.

**Comment by MartinHjelmare** [comment]
I don't think we should drop the subscription approach. We can still use a data coordinator per device for manual updates from the subscription and leverage the coordinator for combining subscription and polling as needed. See eg the Shelly integration.

This will also allow more than one entity per subscription/coordinator/device.

**Reply by ggravlingen** [reply]
Not sure I understand how that should be done exactly, but it could probably be worked out if we decide for the coordinator route.

An objection I might have to subscribing to each device's events is that I believe subscriptions to be the reason for overloading the gateway and thus causing it to stop working. A polling approach could possibly reduce load and make the gateway act in a more stable manner.

Made the repo public, this is how I ended up building it while finding my way around how the coordinator works: https://github.com/ggravlingen/tradfri_custom_component/blob/b441c1c7607ff09916329004e9e48ac79ec2100d/coordinator.py#L61

**Reply by ggravlingen** [reply]
(I did check out the Shelly integration but am having a hard time to follow what's going on there.)

**Reply by ggravlingen** [reply]
I’ve been running the mentioned code as a custom component for the past few days. The gateway keeps crashing in a way I’ve never experienced before. My preliminary conclusion is that the gateway doesn’t like polling in the way I’ve implemented it, and that we should indeed preserve the observe-structure.

**Comment by Mariusthvdb** [comment]
as I have the working version running, please let me add this: both integrations (core and custom) installed on 2 separate systems ofc dont play well simultaneously on the same Tradfri Gateway. 

Since the latest Tradfri firmware update ([1.19 dec 13th](https://ww8.ikea.com/ikeahomesmart/releasenotes/releasenotes.html)), the core integration has become much more stable, not seeing the frequent disconnects anymore.

Installing the Custom integration (to have cover battery and air sensors), and using the coordinator, breaks the core integration. I would be an advocate of a quick turnover, so we can thoroughly test and improve the new integration as soon as a dev version would be available in HA dev.

thanks to @ggravlingen for making this possible in the first place.
ready to test whatever might be needed to get this going.

would be really awesome if could get Tradfri up to par again.

**Comment by ggravlingen** [comment]
Update: I have done some refactoring and there is now support for subscriptions, although I will probably need some help to figure out a more clever way to execute the update. I've been running this stably in my production environment for the past few days as a proof-of-concept.

**Reply by MartinHjelmare** [reply]
I'd create one coordinator per device. When a device updates we want the entities that belong to that device to update and no other entities to update.

The coordinator output data can simply be the device. The platforms can calculate what entities to add for each device and the entities will access the correct device attribute when they are triggered to update by the coordinator.

Use this coordinator method to set the updated data that is triggered by the observation callback:
https://github.com/home-assistant/core/blob/e03283292be22c9188de9ed6ecd973fd99e2857d/homeassistant/helpers/update_coordinator.py#L267-L269

**Reply by ggravlingen** [reply]
Thanks, I will look into this idea.

**Reply by ggravlingen** [reply]
I've implemented this idea and pushed to my repo. Will run this in my production setup for a few days to see if it runs smoothly. If it does, I'm happy to submit a PR if we can decide to go down this route?

**Comment by ggravlingen** [comment, marked as answer]
@MartinHjelmare my custom component with the coordinator-related changes is running stably in my production environment. Would you approve that I turn this discussion into a PR against core? In that case, this is my plan:
- Open a PR with the proposed changes to `coordinator.py`, `cover.py`, `__init__.py`, `const.py` and `base_class.py`.
- Discuss these proposed changes in a code review.
- Once we have resolved all issues in the the files in the first point of this list I will add to the PR the other platforms one by one. This would include battery sensors for the blinds and air quality sensors for the air purifier. Perhaps I should add these sensors in another PR to make the review easier?

**Reply by MartinHjelmare** [reply]
In the first PR, we should just do the changes required to introduce the coordinator, nothing more. Not add any new features, like more sensors etc. 👍

**Reply by ggravlingen** [reply]
Understood!  ✅
